src/compiler/core.ts

```diff
@@ -944,16 +944,54 @@ export function sortAndDeduplicate<T>(array: readonly T[], comparer?: Comparer<T
 /** @internal */
 export function arrayIsSorted<T>(array: readonly T[], comparer: Comparer<T>) {
     if (array.length < 2) return true;
-    let prevElement = array[0];
-    for (const element of array.slice(1)) {
-        if (comparer(prevElement, element) === Comparison.GreaterThan) {
+    for (let i = 1, len = array.length; i < len; i++) {
+        if (comparer(array[i - 1], array[i]) === Comparison.GreaterThan) {
             return false;
         }
-        prevElement = element;
     }
     return true;
 }
 
+/** @internal */
+export const enum SortKind {
+    None = 0,
+    CaseSensitive = 1 << 0,
+    CaseInsensitive = 1 << 1,
+    Both = CaseSensitive | CaseInsensitive,
+}
+
+/** @internal */
+export function detectSortCaseSensitivity(array: readonly string[], useEslintOrdering?: boolean): SortKind;
+/** @internal */
+export function detectSortCaseSensitivity<T>(array: readonly T[], useEslintOrdering: boolean, getString: (element: T) => string): SortKind;
+/** @internal */
+export function detectSortCaseSensitivity<T>(array: readonly T[], useEslintOrdering: boolean, getString?: (element: T) => string): SortKind {
+    let kind = SortKind.Both;
+    if (array.length < 2) return kind;
+    const caseSensitiveComparer = getString
+        ? (a: T, b: T) => compareStringsCaseSensitive(getString(a), getString(b))
+        : compareStringsCaseSensitive as (a: T | undefined, b: T | undefined) => Comparison;
+    const compareCaseInsensitive = useEslintOrdering ? compareStringsCaseInsensitiveEslintCompatible : compareStringsCaseInsensitive;
+    const caseInsensitiveComparer = getString
+        ? (a: T, b: T) => compareCaseInsensitive(getString(a), getString(b))
+        : compareCaseInsensitive as (a: T | undefined, b: T | undefined) => Comparison;
+    for (let i = 1, len = array.length; i < len; i++) {
+        const prevElement = array[i - 1];
+        const element = array[i];
+        if (caseSensitiveComparer(prevElement, element) === Comparison.GreaterThan) {

```

Comment:
It would be more obvious to me what was going on if this was something like `!== Comparison.LessThanOrEqualTo`

Reply:
“less than or equal to” can’t be represented in a single value the way our `Comparison` system works:

```ts
export const enum Comparison {
    LessThan    = -1,
    EqualTo     = 0,
    GreaterThan = 1
}
```

Reply:
Yeah - just an idle sort of "it would be nice".